src/lib.rs

```diff
@@ -1054,49 +1083,59 @@ fn decode_gzip(data: Vec<u8>) -> Result<Vec<u8>, TiledError> {
     Ok(data)
 }
 
-fn decode_csv<R: Read>(parser: &mut EventReader<R>) -> Result<Vec<Vec<u32>>, TiledError> {
+fn decode_csv<R: Read>(
+    parser: &mut EventReader<R>,
+    width: u32,
+) -> Result<TilesContainer, TiledError> {
     loop {
         match try!(parser.next().map_err(TiledError::XmlDecodingError)) {
             XmlEvent::Characters(s) => {
-                let mut rows: Vec<Vec<u32>> = Vec::new();
-                for row in s.split('\n') {
-                    if row.trim() == "" {
-                        continue;
-                    }
-                    rows.push(
-                        row.split(',')
-                            .filter(|v| v.trim() != "")
-                            .map(|v| v.replace('\r', "").parse().unwrap())
-                            .collect(),
-                    );
-                }
-                return Ok(rows);
+                let data: Vec<u32> = s
+                    .replace('\r', "")
+                    .replace('\n', "")
+                    .split(",")
+                    .map(|v| v.parse().unwrap())
+                    .collect();
+                let number_of_rows = data.len() as u32 / width;
+                let tiles = TilesContainer {
+                    data,
+                    tiles_per_row: width,
+                    number_of_rows,
+                };
+                return Ok(tiles);
             }
             XmlEvent::EndElement { name, .. } => {
                 if name.local_name == "data" {
-                    return Ok(Vec::new());
+                    return Ok(TilesContainer {
```

Comment:
If we're going to return a null container I think it would be better to return `None` (or some kind of `Error`) rather than an empty tiles container.

Reply:
I guess that depends on whether it's actually an error. If there is a 0x0 layer saved somehow, this null container would be the expected return value. You can't create that in Tiled, though (it enforces maps to be 1x1). Might be more useful to raise an error when the dimensions of the returned tile container are not the expected ones?

Reply:
Well, if it's not possible to have a 0x0 layer, then I guess it could be interpreted as an error. If this branch is normally inaccessible with any correct Tiled map then maybe it should return `InvalidData` or `Corrupted` or something like that as an error.

Reply:
It's possible to create such a file outside of Tiled and it will load such a file fine. I think it's not useful to consider it an error. It is an error though, if the dimensions of the layer data don't match up with the `width` and `height` attributes. Hence I would suggest a check at a higher level.